### DISCUSSION TITLE
Connections for ReadFrom settings

### DISCUSSION THREAD

**Discussion by akshagr** [redis/lettuce#2060] [Q&A] [answered]
Hi @mp911de

I do have a Redis cluster with 1 shard (1 master, 1 Replica). I have created a RedisClusterClient, and StatefulRedisClusterConnection. I am using ReadFrom settings on connections, have some doubts -

1) While using ReadFrom.MASTER on connection, and we are passing read, write calls over this connection. Will it create TCP connection on Replica node too ?
2)  While using ReadFrom.MASTER on connection, what if connection is exhausted with the read calls, will it create connection to Replica and pass calls to it ?
3) While using ReadFrom.REPLICA_PREFERRED on connection, will it create connection to Master, if we are passing read calls only over this connection ?

Thanks

**Comment by mp911de** [comment, marked as answer]
`ReadFrom.MASTER` uses internally an optimization to the write-connection only. Since replicas aren't master nodes, `ReadFrom.MASTER` does not create read connections to replicas. You might see connections to replicas only in two cases:

1. Topology update
2. Topology changes when a replica is being promoted to a `MASTER` and if a `MASTER` is demoted to `REPLICA`, then the connection remains still open.

If a master is down, then the command is going to timeout. If a slot is not covered, then you will experience a `PartitionSelectorException` exception.

About `REPLICA_PREFERRED`: `PREFERRED` modes prefer to use a particular type of node. We always check if the connection is open and try to use a healthy connection during node selection. Any other `ReadFrom` than `UPSTREAM` leads to additional connection creation that are configured with the `READONLY` flag to avoid cluster redirects when issuing read commands to replicas.

**Reply by akshagr** [reply]
Thanks a lot for answer !!
